### DISCUSSION TITLE
Add filterNonNullish

### DISCUSSION THREAD

**Discussion by david-shortman** [ReactiveX/rxjs#7393] [Ideas / Feature request]
It's common to filter non-nullish values. Today, in Typescript, that would be written as:

```ts
// Accessing the type
filter((value): value is ValueType => value !== null && value !== undefined)

// Avoiding accessing the type
filter((value): value is NonNullable<typeof value> => value !== null && value !== undefined)
```

Instead, it would be preferable to have an operator like:
```ts
filterNonNullish()
```

Existing alternatives include

Defining a custom function:
```ts
const nonNullish = <T>(t: T | null | undefined): t is T => t !== null && t !== undefined;
filter(nonNullish)
```

Using `Boolean` (but this checks for falsy-ness, which is less strict/ more error prone):
```ts
filter(Boolean)
```

**Comment by benlesh** [comment]
I don't think we'd add that directly to RxJS, especially when you can have a relatively straightforward utility function that is more generally useful:

```ts
function isNonNull<T>(value: T): value is NonNullable<T> {
  return value != null;
}

declare const source$: Observable<false | 0 | '' | null | undefined>;

const result$ = source$.filter(isNonNull); // Observable<false | 0 | ''>;
```

Protip, if you don't like seeing `NonNullable<whatever>` in your IDE Tooltips, you can just use what it's shorthand for: `T & {}`.

And if you wanted to have your own custom operator, you could do this... (I've changed the name to `filterOutNonNullish` because `filter` _allows_ values when the predicate passes, and `filterNonNullish` is ambiguous IMO.

```ts
export const filterOutNonNullish = filter(isNonNull);

// usage

source$.pipe(
  filterOutNonNullish
)
```
